### PR TITLE
NTR - Improve testing guides

### PR DESCRIPTION
**Pull request by leichteckig**

### CODE REVIEW

File: resources/guidelines/testing/e2e-best-practises.md [hunk 1]
```diff
@@ -1,52 +1,53 @@
...
+To sum it up briefly for your End-to-End testing context, End-Tto-End tests are the one tests being slow and thus expensive to maintain. That's why we need a way to prioritize our test cases.
 
-  stories \(workflows\) that contain many components and views. 
+### When should I write an end-to-end test?
 
-* Sometimes unit tests are suited better: For example, use an E2E test to test your application's reaction on a 
-
-  failed validation, not the validation itself.
+You see, due to running times it is not advisable to cover every single workflow available. The following criteria may help you with that:
+* Cover the most general, most used workflows of a feature, e.g. CRUD operations. The term "happy path" describes those workflows quite well.
```
Comment (lernhart) [inline resources/guidelines/testing/e2e-best-practises.md:24]: Maybe link to Wiki or maybe there is some better source for `happy path` ? 

The term was new to me

File: resources/guidelines/testing/e2e-best-practises.md [hunk 2]
```diff
@@ -1,52 +1,53 @@
...
 
 You need to focus on the workflow to be tested to ensure minimum test runtimes and to get a valid result of your test case if it fails. Fot this workflow, you have to think like the end-user would do: Focus on usage of your feature, not technical implementation.
 
 Other examples of steps or workflow to cut off the actual tests are:
+* The routines which should only provide the data we need: Just use test fixtures to create this data to have everything available before the test starts.
+* Logging in to the Administration: You need it in almost every Administration test, but writing it in all tests is pure redundancy and way more error sensitive.
 
-* The routines which should only provide the data we need: Just use test fixtures to create this data 
+{% hint style="info" %} This [scope practise](https://docs.cypress.io/guides/references/best-practices.html#Organizing-Tests-Logging-In-Controlling-State) is also mentioned in Cypress' best practises as well.
```
Comment (lernhart) [inline resources/guidelines/testing/e2e-best-practises.md:40]:
```suggestion
{% hint style="info" %} This [scope practice](https://docs.cypress.io/guides/references/best-practices.html#Organizing-Tests-Logging-In-Controlling-State) is also mentioned in Cypress' best practices as well.
```

File: resources/guidelines/testing/e2e-best-practises.md [hunk 3]
```diff
@@ -77,3 +78,5 @@ cy.wait('@getData').then((xhr) => {
 });
 ```
 
+{% hint style="info" %} This [best practise](https://docs.cypress.io/guides/references/best-practices#Unnecessary-Waiting) is also mentioned in Cypress' best practises as well. Actually, it can be considered as general best practise to avoid flakiness.
```
Comment (lernhart) [inline resources/guidelines/testing/e2e-best-practises.md:81]:
```suggestion
{% hint style="info" %} This [best practice](https://docs.cypress.io/guides/references/best-practices#Unnecessary-Waiting) is also mentioned in Cypress' best practices as well. Actually, it can be considered as general best practice to avoid flakiness.
```

**Review by lernhart — CHANGES_REQUESTED** [review]
Just some minor things

**Review by jleifeld — CHANGES_REQUESTED** [review]

**Review by lernhart — APPROVED** [review]

File: resources/guidelines/testing/e2e-best-practises.md [hunk 4]
```diff
@@ -77,3 +119,17 @@ cy.wait('@getData').then((xhr) => {
...
+
+## Cypress' commands and their queue
+
+{% hint style="danger" %} Using vanilla JavaScript logic alongside cypress commands without further caution {% endhint %}
+
+{% hint style="success" %} If you need vanilla Javascript in your test, wrap it in a Cypress' `then` or build a custom command in order to get it queued. {% endhint %}
+
+Cypress commands are asynchronous and get queued for execution at a later time. During execution, subjects are yielded from one command to the next, and a lot of helpful Cypress code runs between each command to ensure everything is in order.
+
+This won't happen with Vanilla JS though, it will be executed immediately. In the worst case, this difference can cause timing issues. So always wrap your vannila JavaScript code into Cypress commands or `then` in order to make use of Cypress' command queue.
```
Comment (JoshuaBehrens) [inline resources/guidelines/testing/e2e-best-practises.md:133]: Vannila -> Vanilla

File: resources/guidelines/testing/e2e-best-practises.md [hunk 5]
```diff
@@ -1,59 +1,101 @@
...
+* **Cover the most general, most used workflows of a feature**, e.g. CRUD operations. The term "[happy path](https://en.wikipedia.org/wiki/Happy_path)" describes those workflows quite well.
+* **Beware the critical path**: Cover those workflows with E2E tests, which are most vulnerable and would cause most damage if broken.
+* **Avoid duplicate coverage**: E2E tests should only cover what they can cover, usually big-picture user stories \(workflows\) that contain many components and views.
+  * Sometimes unit tests are suited better: For example, use an E2E test to test your application's reaction on a failed validation, not the validation itself.
+
+## Workflow-based end-to-end tests
+
+{% hint style="danger" %} Write E2E test as you would write unit tests. {% endhint %}
+
+{% hint style="success" %} Write E2E tests in a "workflow-based" manner, that means writing the test describing a reals
```
Comment (JoshuaBehrens) [inline resources/guidelines/testing/e2e-best-practises.md:43]: You got me puzzled here:

I assume 
reals -> real

**Review by JoshuaBehrens — CHANGES_REQUESTED** [review]

File: resources/guidelines/testing/e2e-best-practises.md (hunk 5 above)
Comment (leichteckig) [reply resources/guidelines/testing/e2e-best-practises.md:43]: Ah, bad typo 🙈

File: resources/guidelines/testing/e2e-best-practises.md (hunk 4 above)
Comment (leichteckig) [inline resources/guidelines/testing/e2e-best-practises.md:133]:
```suggestion
This won't happen with Vanilla JS though, it will be executed immediately. In the worst case, this difference can cause timing issues. So always wrap your vanilla JavaScript code into Cypress commands or `then` in order to make use of Cypress' command queue.
```

File: resources/guidelines/testing/e2e-best-practises.md (hunk 5 above)
Comment (leichteckig) [inline resources/guidelines/testing/e2e-best-practises.md:43]:
```suggestion
{% hint style="success" %} Write E2E tests in a "workflow-based" manner, that means writing the test describing a real
```

**Review by JoshuaBehrens — APPROVED** [review]
